### PR TITLE
HAL for dot product added

### PR DESCRIPTION
**Pull request by savuor**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by asmorkalov** [comment]
```
modules/core/src/hal_replacement.hpp:904: trailing whitespace.
+ * 
modules/core/src/hal_replacement.hpp:913: trailing whitespace.
+ * @return int 
```

**Review by asmorkalov — APPROVED** [review]

File: modules/core/src/hal_replacement.hpp [hunk 1]
```diff
@@ -899,6 +899,27 @@ inline int hal_ni_meanStdDev(const uchar* src_data, size_t src_step, int width,
...
+ * @param b_data Pointer to 1st 2nd image data
+ * @param b_step Stride of 1st 2nd image
+ * @param width Width of both images
+ * @param height Height of both images
+ * @param type Data type of both images, for example CV_8U or CV_32F
+ * @param dot_val Pointer to resulting dot product value
+ * @return int
+ */
+inline int hal_ni_dotProduct(const uchar* a_data, size_t a_step, const uchar* b_data, size_t b_step, int width, int height,
+                             int type, double *dot_val)
```
Comment (vpisarev) [inline modules/core/src/hal_replacement.hpp:916]: can we return int64_t instead of double?

Comment (asmorkalov) [reply modules/core/src/hal_replacement.hpp:916]: Public interface: `double Mat::dot(InputArray _mat) const`. `int64_t` looks strange. Also the `::dot()` may handle floats and doubles.

**Review by vpisarev — APPROVED** [review]

File: modules/core/src/matmul.dispatch.cpp [hunk 2]
```diff
@@ -995,9 +995,18 @@ double Mat::dot(InputArray _mat) const
     CV_INSTRUMENT_REGION();
 
     Mat mat = _mat.getMat();
+    CV_Assert_N( mat.type() == type(), mat.size == size);
```
Comment (opencv-alalek) [inline modules/core/src/matmul.dispatch.cpp:998]: `CV_Assert_N` is deprecated due to malformed or useless error messages.
Use `CV_Check*` instead.

**Pull request merged**